tests/cnf/ran/internal/ranconfig/config.go

```diff
@@ -20,20 +25,49 @@ const (
 // RANConfig contains configuration for the RAN directory.
 type RANConfig struct {
 	*cnfconfig.CNFConfig
+	*HubConfig
+	*Spoke1Config
+	*Spoke2Config
+
 	MetricSamplingInterval string   `yaml:"metricSamplingInterval" envconfig:"ECO_CNF_RAN_METRIC_SAMPLING_INTERVAL"`
 	NoWorkloadDuration     string   `yaml:"noWorkloadDuration" envconfig:"ECO_CNF_RAN_NO_WORKLOAD_DURATION"`
 	WorkloadDuration       string   `yaml:"workloadDuration" envconfig:"ECO_CNF_RAN_WORKLOAD_DURATION"`
 	StressngTestImage      string   `yaml:"stressngTestImage" envconfig:"ECO_CNF_RAN_STRESSNG_TEST_IMAGE"`
 	CnfTestImage           string   `yaml:"cnfTestImage" envconfig:"ECO_CNF_RAN_TEST_IMAGE"`
-	HubKubeconfig          string   `envconfig:"ECO_CNF_RAN_KUBECONFIG_HUB"`
-	Spoke2Kubeconfig       string   `envconfig:"ECO_CNF_RAN_KUBECONFIG_SPOKE2"`
-	BmcUsername            string   `yaml:"bmcUsername" envconfig:"ECO_CNF_RAN_BMC_USERNAME"`
-	BmcPassword            string   `yaml:"bmcPassword" envconfig:"ECO_CNF_RAN_BMC_PASSWORD"`
-	BmcHosts               string   `yaml:"bmcHosts" envconfig:"ECO_CNF_RAN_BMC_HOSTS"`
-	BmcTimeout             string   `yaml:"bmcTimeout" envconfig:"ECO_CNF_RAN_BMC_TIMEOUT"`
 	OcpUpgradeUpstreamURL  string   `yaml:"ocpUpgradeUpstreamUrl" envconfig:"ECO_CNF_RAN_OCP_UPGRADE_UPSTREAM_URL"`
 	PtpOperatorNamespace   string   `yaml:"ptpOperatorNamespace" envconfig:"ECO_CNF_RAN_PTP_OPERATOR_NAMESPACE"`
 	TalmPreCachePolicies   []string `yaml:"talmPreCachePolicies" envconfig:"ECO_CNF_RAN_TALM_PRECACHE_POLICIES"`
+	ZtpSiteGenerateImage   string   `yaml:"ztpSiteGenerateImage" envconfig:"ECO_CNF_RAN_ZTP_SITE_GENERATE_IMAGE"`
+}
+
+// HubConfig contains the configuration for the hub cluster, if present.
+type HubConfig struct {
+	HubAPIClient  *clients.Settings
+	HubOCPVersion string
+	ACMVersion    string
+	TALMVersion   string
+	ZTPVersion    string
+	HubKubeconfig string `envconfig:"ECO_CNF_RAN_KUBECONFIG_HUB"`
+}
+
+// Spoke1Config contains the configuration for the spoke 1 cluster, which should always be present.
+type Spoke1Config struct {
```

Comment:
I think we can use only one struct and one constructor func newSpokeConfig that returns SpokeConfig? e.g., type SpokeConfig struct {...}, then var Spoke1Config SpokeConfig . 

func newSpokeConfig (configFile string) SpokeConfig {...}
If the config file does not contain certain field, then just set them to empty/nil.   

Spoke1Config = newSpokeConfig(spoke1ConfigFile)
Spoke2Config = newSpokeConfig(spoke2ConfigFile)  # if spoke2 exists

And for SpokeConfig and HubConfig struct, I think we can combine the day2 operator versions into one map field. So it is easier to get all relevant operator versions in one function call.

e.g., 
```
type HubConfig struct {
	APIClient  *clients.Settings
	OCPVersion string
	Day2OperatorVersions map[string]string
        ZTPVersion string
	Kubeconfig string `envconfig:"ECO_CNF_RAN_KUBECONFIG_HUB"`
}

type SppkeConfig struct {
	BMC        *bmc.BMC
	APIClient  *clients.Settings
	ClusterName       string
	OCPVersion string
        Day2OperatorVersions map[string]string
        Kubeconfig string
	BMCUsername      string 
        BMCPassword      string 
	BMCHosts         []string 
	BMCTimeout       time.Duration `yaml:"bmcTimeout" envconfig:"ECO_CNF_RAN_BMC_TIMEOUT"`
}

```

func GetOperatorVersionsFromCsv(client *clients.Settings, operatorNames map[string]string) map[string]string {...}

operatorNames uses namespace as key and name as value.
return value can be a map with operator name as key and version as value.

Alternatively, OperatorInfo could be a struct itself, and both input can be a list of OperatorInfo . I think this method will require more code and not sure if it will give us much benefit since at this top level we will very likely never need to collect/store more info other than name, namespace and version for an operator. 

For spoke operators, we will need to collect ptp, sriov, logging, local-storage, and lifecycle-agent versions. And LCA might not be installed in all envs.

Let me know how you think about this.

Reply:
@yliu127 I considered using a single SpokeConfig struct before but ultimately decided against it since it prevents using envconfig with struct tags properly. Instead, we would have to read (and parse) different environment variables depending on which spoke config is being loaded. Also, Spoke 2 is only used by TALM and never for BMC so it doesn't seem unreasonable to consider it a different type unless it starts being used more similarly to Spoke 1.

I like the idea for collecting spoke operator versions, though we won't need them until the PTP suite gets migrated so maybe we can add them to the config during that suite?

Reply:
@klaskosk thanks for clarifying. Let's just do the day2 operator change for hub first? We need acm, mce, gitops, and talm for hub. Although in the ztp/talm suite, we only need to use ztp/talm version to determine whether a test is valid, but when test fails, we often need to provide versions for all relevant components.